### PR TITLE
Feature/31 continuous cm estimation

### LINKED ISSUE #133: thrustCMEstimation fsw module
**Describe your use case**
This module reads in torque measurements from the integral feedback control law and implements a sequential, weighted leas-squares solver to triangulate the location of the CM.

**Describe alternatives solutions you've considered**
Alternatives solutions consist in not estimating the CM location and relying on inaccurate knowledge of the CM location for a very long period of time, until a proper calibration maneuver can be performed. However, this can lead to momentum injection into the system, which could ultimately cause reaction wheel saturation and loss of attitude.

### PR DESCRIPTION
**Pull request by rcalaon**
* **Tickets addressed:** bsk-031
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
This module implements a sequential, weighted leas-squares algorithm to process torque measurements and triangulate the location of the center of mass, relying on the knowledge of the thrust vector and application point. At least two measurements are required to triangulate the location. The estimated CM location allows to stabilize the thrust vector in order to continuously dump momentum even in the presence of external disturbances such as SRP torque. Commit 1 adds a line to the cmake file to build the module in a new folder. Commit 2 adds a message containing estimated data and residuals. Commit 3 contains the initial module files. Commit 4 checks the connected messages. Commit 5 performs the actual LS estimation. Commits 6 and 7 contain the unit test and documentation, respectively.

## Verification
A Unit Test is provided. Because this is an estimation, no exact result can be verified. However, this test checks the correctness of the results based on consideration regarding the estimated state error and covariance.

## Documentation
Documentation is provided to explain how the estimator works.

## Future work
The estimator can be expanded to include the unmodeled perturbation in the state, in order to properly estimate the CM and the unmodeled perturbation itself.

### CODE REVIEW

File: src/architecture/msgPayloadDefC/CMEstDataMsgPayload.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+ ISC License
+
+ Copyright (c) 2016, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
```
Comment (patkenneally) [inline src/architecture/msgPayloadDefC/CMEstDataMsgPayload.h]: :large_orange_diamond: I think this needs to be "Copyright (c) 2023 Laboratory for Atmospheric and Space Physics, University of Co...." , as you have in your other new files.

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+#include <string.h>
+#include <array>
+#include <math.h>
+
+class ThrustCMEstimation: public SysModel {
+public:
+    ThrustCMEstimation();
+    ~ThrustCMEstimation() override;
+    void SelfInit();                   //!< Self initialization for C-wrapped messages
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h]: :large_orange_diamond: You can mark `SelfInit()` as overridden and delete the redundant comment.

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+
+/*! @brief Top level structure for the thrust CM estimation kalman filter.
+ Used to estimate the spacecraft's center of mass position with respect to the B frame.
+ */
+
+#ifndef THRUSTCMESTIMATION_H
+#define THRUSTCMESTIMATION_H
+
+#include "architecture/_GeneralModuleFiles/sys_model.h"
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h]: :large_blue_circle: If you want bonus points 😄 , but you can consider it for future code. you can order your includes by the following and then alphabetically. 
1.     h file corresponding to this cpp file (if applicable)
2.     headers from the same component,
3.     headers from other components,
4.     system headers.

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+#  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+#  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+#  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+#
+#
+#   Unit Test Script
+#   Module Name:        thrustCMEstimation
+#   Author:             Riccardo Calaon
+#   Creation Date:      August 28, 2023
+#
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: :small_red_triangle_down: Delete this. The Git history contains this information.

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    ReadFunctor<CmdTorqueBodyMsgPayload>    intFeedbackTorqueInMsg;
+    ReadFunctor<AttGuidMsgPayload>          attGuidInMsg;
+    ReadFunctor<VehicleConfigMsgPayload>    vehConfigInMsg;
+    Message<CMEstDataMsgPayload>            cmEstDataOutMsg;
+    Message<VehicleConfigMsgPayload>        vehConfigOutMsg;
+    VehicleConfigMsg_C                      vehConfigOutMsgC = {};
+
+    Eigen::Vector3d r_CB_B;
+    Eigen::Vector3d P0;
+    Eigen::Vector3d R0;
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h]: :small_red_triangle_down: I think these three member variables should be `private` and passed as parameters to the module's constructor. They are initial values which are used in reset. I think they shouldn't be settable thoughout a sim. Rather, if a user wants to change these they should create a new module. Happy to chat more about this and hear your thoughts.

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.cpp [hunk 6]
```diff
@@ -72,5 +72,90 @@ void ThrustCMEstimation::Reset(uint64_t CurrentSimNanos)
...
+
+    Eigen::Vector3d y;       // measurement
+    Eigen::Vector3d preFit;  // pre-fit residual
+    Eigen::Vector3d postFit; // post-fit residual
+    Eigen::Matrix3d H;       // linear model
+    Eigen::Matrix3d K;       // kalman gain
+    Eigen::Matrix3d S;
+
+    /*! assign preFit and postFit residuals to NaN, rewrite them in case of measurement update */
+    preFit  = {nan("1"), nan("1"), nan("1")};
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.cpp:104]: :large_orange_diamond: Can you help me understand why a nan is chosen here as opposed to something else? I don't have a strong opinion and so would like to hear yours. We can huddle about this one.

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.cpp [hunk 7]
```diff
@@ -72,5 +72,90 @@ void ThrustCMEstimation::Reset(uint64_t CurrentSimNanos)
...
+        /*! measurement covariance update  */
+        this->P = (this->I - K * H) * this->P;
+
+        /*! post-fit residuals */
+        postFit = y - H * this->r_CB_est;
+    }
+
+    /*! compute state 1-sigma covariance */
+    Eigen::Vector3d sigma;
+    for (int i=0; i<3; i++) {
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.cpp]: :large_blue_circle: A fun thing to note; you could pre increment here, and any other for loop, and save two extra machine instructions (when the loop is compiled). A good explanation of why that works

https://stackoverflow.com/a/4706225

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+#  ISC License
+#
+#  Copyright (c) 2023, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: :large_orange_diamond: Switch to LASP

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        uMeasNoise = np.random.multivariate_normal([0,0,0], R0, size=1)
+        intTorque.torqueRequestBody = -np.cross(r_TB_B[i], T_B[i]) + uMeasNoise[0]
+        intFeedbackTorqueMsg.write(intTorque)
+
+        t += dT
+        unitTestSim.ExecuteSimulation()
+        unitTestSim.ConfigureStopTime(macros.sec2nano(t-0.1))
+
+    #   retrieve the logged data
+    state = cmEstimateLog.state
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: :large_orange_diamond: unused variable

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    stateErr = cmEstimateLog.stateError
+    sigma = cmEstimateLog.covariance
+    preFit = cmEstimateLog.preFitRes
+    postFit = cmEstimateLog.postFitRes
+    timeData = cmEstimateLog.times() * macros.NANO2SEC
```
Comment (patkenneally) [inline src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: :large_orange_diamond: unused variable

**Review by patkenneally — CHANGES_REQUESTED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

Let's have a quick chat and then this is good to pull in with minor adjustments. Nicely done!

File: src/architecture/msgPayloadDefC/CMEstDataMsgPayload.h (hunk 1 above)
Comment (rcalaon) [reply src/architecture/msgPayloadDefC/CMEstDataMsgPayload.h]: Done

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h (hunk 3 above)
Comment (rcalaon) [reply src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h]: I live for them bonus points. I tried!

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py (hunk 4 above)
Comment (rcalaon) [reply src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: Done

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py (hunk 8 above)
Comment (rcalaon) [reply src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: Done

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py (hunk 9 above)
Comment (rcalaon) [reply src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: Removed

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py (hunk 10 above)
Comment (rcalaon) [reply src/fswAlgorithms/stateEstimation/thrustCMEstimation/_UnitTest/test_thrustCMEstimation.py]: Removed

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h (hunk 2 above)
Comment (rcalaon) [reply src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.h]: Done

File: src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.cpp (hunk 7 above)
Comment (rcalaon) [reply src/fswAlgorithms/stateEstimation/thrustCMEstimation/thrustCMEstimation.cpp]: So I just change "i++" to "++i"? Done

**Review by patkenneally — APPROVED** [review]

**Pull request merged**